crates/muvm/src/guest/net.rs

```diff
@@ -1,14 +1,224 @@
+        .ifi_family(RtAddrFamily::Unspecified)
+        .ifi_type(Arphrd::Ether).ifi_index(2)
+        .ifi_change(mask).ifi_flags(set)
+        .build()?;
+
+    let _: NlRouterReceiverHandle<Rtm, Ifinfomsg> =
+        rtnl.send(Rtm::Newlink, NlmF::REQUEST, NlPayload::Payload(ifinfomsg))?;
+
+    Ok(())
+}
+
+/// Add or delete IPv4 routes for eth0 (interface index 2)
+fn route4_eth0(rtnl: &NlRouter, what: Rtm, gw: Ipv4Addr) -> Result<()> {
+    let rtmsg = RtmsgBuilder::default()
+        .rtm_family(RtAddrFamily::Inet)
+        .rtm_dst_len(0).rtm_src_len(0).rtm_tos(0)
+        .rtm_table(RtTable::Main).rtm_protocol(Rtprot::Boot)
+        .rtm_scope(RtScope::Universe).rtm_type(Rtn::Unicast)
+        .rtm_flags(RtmF::empty())
+        .rtattrs(RtBuffer::from_iter([
+            RtattrBuilder::default()
+                .rta_type(Rta::Oif)
+                .rta_payload(2)
+                .build()?,
+            RtattrBuilder::default()
+                .rta_type(Rta::Dst)
+                .rta_payload(Ipv4Addr::UNSPECIFIED.octets().to_vec())
+                .build()?,
+            RtattrBuilder::default()
+                .rta_type(Rta::Gateway)
+                .rta_payload(gw.octets().to_vec())
+                .build()?
+        ]))
+        .build()?;
+
+    let _: NlRouterReceiverHandle<Rtm, Rtmsg> =
+        rtnl.send(what, NlmF::CREATE | NlmF::REQUEST,
+                  NlPayload::Payload(rtmsg))?;
+
+    Ok(())
+}
+
+/// Add or delete IPv4 addresses for eth0 (interface index 2)
+fn addr4_eth0(rtnl: &NlRouter, what: Rtm, addr: Ipv4Addr, prefix_len: u8)
+                       -> Result<()> {
+    let ifaddrmsg = IfaddrmsgBuilder::default()
+        .ifa_family(RtAddrFamily::Inet)
+        .ifa_prefixlen(prefix_len)
+        .ifa_scope(RtScope::Universe)
+        .ifa_index(2)
+        .rtattrs(RtBuffer::from_iter([
+            RtattrBuilder::default()
+                .rta_type(Ifa::Local)
+                .rta_payload(addr.octets().to_vec())
+                .build()?,
+            RtattrBuilder::default()
+                .rta_type(Ifa::Address)
+                .rta_payload(addr.octets().to_vec())
+                .build()?,
+        ]))
+        .build()?;
+
+    let _: NlRouterReceiverHandle<Rtm, Ifaddrmsg> =
+        rtnl.send(what, NlmF::CREATE | NlmF::REQUEST,
+                  NlPayload::Payload(ifaddrmsg))?;
+
+    Ok(())
+}
+
+/// Send DISCOVER with Rapid Commit, process ACK, configure address and route
+fn do_dhcp(rtnl: &NlRouter) -> Result<()> {
+    // Temporary link-local address and route avoid the need for raw sockets
+    route4_eth0(rtnl, Rtm::Newroute, Ipv4Addr::UNSPECIFIED)?;
+    addr4_eth0(rtnl, Rtm::Newaddr, Ipv4Addr::new(169, 254, 1, 1), 16)?;
+
+    // Send request (DHCPDISCOVER)
+    let socket = UdpSocket::bind("0.0.0.0:68").expect("Failed to bind");
+    let mut buf = [0; 576 /* RFC 2131, Section 2 */ ];
+
+    const REQUEST: [u8; 300 /* From RFC 951: >= 60 B of options */ ] = [
```

Comment:
I have a suggestion: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b68067b21b871a7d80e6d9a81d564e92

Reply:
How would you initialise bytes after the `sname` and `file` fields, though? That is, the DHCP cookie and the options.

Besides, if you prefer a version with everything on its own line, a struct with `#[repr(packed)]` could be used instead (at least up to the options... those would need a different description). [This](https://passt.top/passt/tree/dhcp.c?id=020c8b7127e38872e68bffb30ad388001e088552#n86), for example, is how passt does it in C.

But my idea here was, given that we we don't have to be flexible at all, to keep it in sixteen lines (because it's possible), so that you have an immediate overview of the whole thing without scrolling back and forth.

Reply:
I consider the struct to be an overkill like you said, but I also think the original formatting you had was unreadable because they're all packed together. I think my suggestion allows easily seeing what's at byte N.

Reply:
But still, I don't understand: how would you initialise the cookie and option bytes? That is, these ones:

```
        0x63, 0x82, 0x53, 0x63, /* DHCP (magic) cookie, then options: */
        53, 1, 1 /* DISCOVER */, 80, 0 /* Rapid commit */, 0xff, // Done
```

...they come after a lot of zeroes. At the beginning, I tried something similar to your suggestion, but instead of ending it with `r`, I tried to include a given number of zeroes with [0u8; n] in the middle of the initialiser. But it's not supported, and nothing like that seems to be supported from what I read.

> I think my suggestion allows easily seeing what's at byte N.

It's actually not so important for DHCP, because fields are not aligned by design. Making it a bit artistic and all squared looked easier on my eyes to me, which makes me wonder: can it be that our editors have different setting so what's readable for me looks messy to you? This is how it looks like for me, in the editors I use:

![geany](https://github.com/user-attachments/assets/4fa4d8d0-7f96-4ad5-8360-43a06e32852f)

![joe](https://github.com/user-attachments/assets/62aa3b47-e42c-4b9b-827d-55dc69be3220)

Reply:
Readability is subjective, of course. A sea of zeroes interspersed with some values and comments is hard to follow IMO.

I have some ideas on what could work. Will give it a try later.

Reply:
> A sea of zeroes interspersed with some values and comments is hard to follow IMO.

but... but... art :cry:

Reply:
There's [`concat_bytes!`](https://doc.rust-lang.org/std/macro.concat_bytes.html) but it's still unstable: https://github.com/rust-lang/rust/issues/87555

I think we can use [`const_str::concat_bytes!`](https://docs.rs/const-str/latest/const_str/macro.concat_bytes.html) in the meantime.

Reply:
https://github.com/sbrivio-rh/muvm/pull/1

Reply:
I really appreciate the effort, and the creative usage of `concat_bytes!`, I just can't refrain from noticing that a compact, four-byte aligned, sixteen-line representation of a simple UDP datagram is now thirty-six lines, more than double, with a lot of RFC-951-derived cruft (which nobody cares about) spelled out prominently.

It's harder to get a quick overview of the options (the only thing that really matters here).

> Readability is subjective, of course

...except that your way is better? I don't get it.

Well, anyway, if `cargo fmt` is enforced, or in any case if this is the only alternative to avoid screen-scraping from passt (and getting the thing to start fast), so be it.

Where do I click? Should I accept your pull request on my tree and then force-push here? I have to admit that while I've been using git for 19 years, I'm kind of lost with GitHub. I would need to update the comments, too.

By the way, is there a way to calculate the 54 at the end using something equivalent to C's `300 - offsetof(struct msg, end)`?

Reply:
I'll try to use line-end comments instead if that's preferable.

> By the way, is there a way to calculate the 54 at the end using something equivalent to C's 300 - offsetof(struct msg, end)?

I'll define `REQUEST_LEN` :)

Reply:
> ...except that your way is better? I don't get it.

I'm sorry that it came off that way. What I meant is that there's no way for us to judge what's the right way here, except that yes, the formatting you had definitely trips up `rustfmt`, and even if we were to add an ignore here, it'd still look out of place with the rest of the codebase...

Reply:
> I'll try to use line-end comments instead if that's preferable.

...I don't think it would make a big difference, I still see it as an explosion of verbosity for the sake of pretending to document details which happen to come from the mid 80s and which are not relevant anymore.

> I'll define `REQUEST_LEN` :)

It's not so much about the "300" itself, it's about the`offsetof` equivalent to calculate the padding. I couldn't find any.

Reply:
I found https://doc.rust-lang.org/core/mem/macro.offset_of.html but would it help in our case, considering we only have an array?

Reply:
Updated. Now 21 lines vs. your original 17 lines. I hope it's not too inelegant? :bow:

Reply:
> I found https://doc.rust-lang.org/core/mem/macro.offset_of.html but would it help in our case, considering we only have an array?

Aaah, right, and turning that into a struct makes it quite verbose again. Never mind, I don't think we'll ever need to change those.

> Updated. Now 21 lines vs. your original 17 lines. I hope it's not too inelegant? 🙇

Thanks, much much better. I still cry a bit inside because of what happened to my square packet, but okay, if `rustfmt` says so...

I'll now carefully click and try to merge things back to here...

Reply:
Not carefully enough, now I have that dangling Cargo.toml update from you... grr. I'm trying to rebuild a sane situation but it's taking a bit.

Reply:
Pushed, I hope it's fine now...

But really, what do you all do with Cargo.toml if you can't `gitignore` it?

Reply:
Do you mean `Cargo.lock`? Just `git add` it after you run `cargo build` and it should be correct.

Reply:
Right, sorry, `Cargo.lock`. So I'm actually supposed to commit it, even it's automatically generated...?

Reply:
Yes, I read that proposal (or is it a recommendation? It's not stated...) (_after_ proposing the `.gitignore` change, oops). I'm still not convinced it's worth the madness as I never had issues of that sort (well, with C, at least). Anyway, pushing those changes as well...

Reply:
RE: art

https://mastodon.social/@richlv/113567941242044603 :crying_cat_face: